blog/2024-03-28-hybrid-search-typescript/index.mdx

```diff
@@ -0,0 +1,305 @@
+
+### Vector Search
+
+Vector search, also known as Semantic Search, uses machine learning to grasp text context, not just keywords. It converts text to numerical vectors, finding matches based on conceptual similarity for accurate, relevant search results.
+
+![Semantic Search Vector Space](./img/semantic-search.png)
+
+
+### Keyword Search
+
+
+Keyword search finds objects with an exact match between the terms in the query and the documents. Weaviate uses the BM25/F function to calculate the relevancy of a document to a query, which uses both the term frequency and inverse document frequency.
+
+![Keyword Search Illustrated](./img/keyword-search.png)
+
+
+### The Best of Both Worlds
+
+
+Hybrid search in Weaviate combines keyword (BM25) and vector search to leverage both exact term matching and semantic context. With either rankedFusion or relativeScoreFusion algorithms, we merge query results. Merging results in this way helps us have a better balance between precision querying in keyword search and the semantic understanding of Vector Search. 
+
+A code snippet of Hybrid Search in Javascript will look something like this. 
+
+```typescript
+const myCollection = client.collections.get('JeopardyQuestion');
+
+const result = await myCollection.query.hybrid('food',{
+ limit: 2,
+ returnProperties: ['question', 'answer'],
+})
+
+console.log(JSON.stringify(result.objects, null, 2));
+```
+
+
+## Optimizing your Hybrid Search
+
+With Weaviates Hybrid Search, we have two ways to adjust the Hybrid Search algorithm. We can adjust the weights by tweaking our search alpha or changing our ranking algorithm.
+
+### Changing alpha
+
+Think of alpha as a measure of how much your hybrid search should lean towards either Vector Search or keyword search where an alpha of 1 is a pure vector search, and an alpha of 0 is a pure keyword search. It is 0.75 by default. 
+
+
+
+### Changing ranking algorithm 
+
+Here is where it gets fun! Weaviate supports two fusion algorithms as of Weaviate v1.20 i.e. rankedFusion and relativeScoreFusion. 
+
+In a nutshell, rankedFusion (the original Hybrid search algorithm in Weaviate) assigns each object a score based on its position in the search results (in descending order), with the total score derived from the sum of these rank-based scores across both vector and keyword searches.
+
+![Ranked Fusion Algorithm](./img/reciprocal-rank.png)
+
+
+Relative Score Fusion on the other hand normalizes the output of the vector and keyword searches so that the highest value becomes 1 and the lowest 0 and everything in between is scaled to match. The total score is then the sum of the normalized individual search scores. Letting us maintain the relevance given by the original keyword/vector search algorithms, while still being able to compare the results and fuse them. As relativeScoreFusion is the newer algorithm, it would most likely be the better choice for most use cases.
```

Comment:
```suggestion
Relative Score Fusion on the other hand normalizes the output of the vector and keyword searches so that the highest value becomes 1 and the lowest 0 and everything in between is scaled to match. The total score is then the sum of the normalized individual search scores. This lets us maintain the relevance given by the original keyword/vector search algorithms, while still being able to compare the results and fuse them. As relativeScoreFusion is the newer algorithm, it might be the better choice for most use cases.
```